File: src/main/java/com/soapdogg/lispInterpreter/asserter/UserDefinedFormalParametersAsserter.java
package com.soapdogg.lispInterpreter.asserter;

import com.soapdogg.lispInterpreter.exceptions.InvalidUserDefinedNameException;
import lombok.AllArgsConstructor;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

@AllArgsConstructor(staticName = "newInstance")
public class UserDefinedFormalParametersAsserter {

    private final Set<String> invalidNamesSet;

    public void assertFormalParameters(
        final List<String> formalParameters
    ) throws InvalidUserDefinedNameException {
        Set<String> formalParametersSet = new HashSet<>(formalParameters);
        if (formalParameters.size() != formalParametersSet.size()) {
            throw new InvalidUserDefinedNameException("Error! Duplicate formal parameter name!\n");
        }
        formalParametersSet.retainAll(invalidNamesSet);
        if (!formalParametersSet.isEmpty()) {
            throw new InvalidUserDefinedNameException("Error! Invalid formal parameter name!\n");
        }
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/asserter/UserDefinedFormalParametersAsserter.kt
package com.soapdogg.lispInterpreter.asserter

import com.soapdogg.lispInterpreter.exceptions.InvalidUserDefinedNameException
import lombok.AllArgsConstructor
import java.util.*

@AllArgsConstructor(staticName = "newInstance")
class UserDefinedFormalParametersAsserter(
    private val invalidNamesSet: Set<String>
) {

    fun assertFormalParameters(
        formalParameters: List<String>
    ) {
        val formalParametersSet: MutableSet<String> = HashSet(formalParameters)
        if (formalParameters.size != formalParametersSet.size) {
            throw InvalidUserDefinedNameException("Error! Duplicate formal parameter name!\n")
        }
        formalParametersSet.retainAll(invalidNamesSet)
        if (formalParametersSet.isNotEmpty()) {
            throw InvalidUserDefinedNameException("Error! Invalid formal parameter name!\n")
        }
    }
}